File: 프로그래머스/2/42586. 기능개발/기능개발.py
from collections import deque

def solution(progresses, speeds):
    answer = []
    
    progress_q = deque(progresses)
    speeds_q = deque(speeds)
    while progress_q:
        count = 0
        while progress_q and progress_q[0] >= 100:
            if progress_q:
                progress_q.popleft()
                speeds_q.popleft()
                count += 1  
        if count > 0:
            answer.append(count)
            
        for i in range(len(progress_q)):
            progress_q[i] += speeds_q[i]   
        
        
    return answer
        
                
            

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        ArrayList<Integer> answer = new ArrayList<>();
        Deque<Integer> progressQueue = new ArrayDeque<>();
        Deque<Integer> speedQueue = new ArrayDeque<>();
        
        for (int i = 0; i < progresses.length; i++) {
            progressQueue.addLast(progresses[i]);
            speedQueue.addLast(speeds[i]);
        }
        
        int idx = 0;
        while (progressQueue.size() != 0) {
            for (int i = idx; i < progresses.length; i++) {
                progresses[i] += speeds[i];
            }
            
            int count = 0;
            while (idx < progresses.length && progresses[idx] >= 100) {
                count += 1;
                idx += 1;
                progressQueue.removeFirst();
                speedQueue.removeFirst();
            }
            
            if (count > 0) answer.add(count);
        }
        
        return answer.stream().mapToInt(n -> n).toArray();
    }
}